sam/pileup.go

```diff
@@ -0,0 +1,362 @@
+		log.Fatal("ERROR: (GoPileup) input sam/bam must be coordinate sorted")
+	}
+	pileChan := make(chan Pile, 1000)
+	go pileup(pileChan, reads, header, includeNoData, readFilters, pileFilters)
+	return pileChan
+}
+
+// pileup generates multiple Pile based on the input reads and sends the Pile when passed.
+func pileup(send chan<- Pile, reads <-chan Sam, header Header, includeNoData bool, readFilters []func(s Sam) bool, pileFilters []func(p Pile) bool) {
+	pb := newPileBuffer(300) // initialize to 2x std read size
+	refmap := chromInfo.SliceToMap(header.Chroms)
+	for read := range reads {
+		if !passesReadFilters(read, readFilters) {
+			continue
+		}
+		pb.sendPassed(read, includeNoData, refmap, send, pileFilters)
+		updatePile(pb, read, refmap)
+	}
+	close(send)
+}
+
+// passesReadFilters returns true if input Sam is true for all functions in filters.
+func passesReadFilters(s Sam, filters []func(s Sam) bool) bool {
+	for _, f := range filters {
+		if !f(s) {
+			return false
+		}
+	}
+	return true
+}
+
+// passesPileFilters returns true if input Sam is true for all functions in filters.
+func passesPileFilters(p Pile, filters []func(p Pile) bool) bool {
+	for _, f := range filters {
+		if !f(p) {
+			return false
+		}
+	}
+	return true
+}
+
+// updatePile updates the input pileBuffer with the data in s
+func updatePile(pb *pileBuffer, s Sam, refmap map[string]chromInfo.ChromInfo) {
+	var seqPos int
+	refPos := s.Pos
+	refidx := refmap[s.RName].Order
+	for i := range s.Cigar {
+		switch s.Cigar[i].Op {
+		case 'M', '=', 'X': // Match
+			addMatch(pb, refidx, refPos, s.Seq[seqPos:seqPos+s.Cigar[i].RunLength])
+			refPos += uint32(s.Cigar[i].RunLength)
+			seqPos += s.Cigar[i].RunLength
+
+		case 'D': // Deletion
+			addDeletion(pb, refidx, refPos, s.Cigar[i].RunLength)
+			refPos += uint32(s.Cigar[i].RunLength)
+
+		case 'I': // Insertion
+			addInsertion(pb, refidx, refPos-1, s.Seq[seqPos:seqPos+s.Cigar[i].RunLength])
+			seqPos += s.Cigar[i].RunLength
+
+		default:
+			if cigar.ConsumesReference(s.Cigar[i].Op) {
+				refPos += uint32(s.Cigar[i].RunLength)
+			}
+			if cigar.ConsumesQuery(s.Cigar[i].Op) {
+				seqPos += s.Cigar[i].RunLength
+			}
+		}
+	}
+}
+
+// addMatch to pileBuffer
+func addMatch(pb *pileBuffer, refidx int, startPos uint32, seq []dna.Base) {
+	for i := range seq {
+		pb.addBase(refidx, startPos+uint32(i), seq[i])
+	}
+}
+
+// addDeletion to pileBuffer
+func addDeletion(pb *pileBuffer, refidx int, startPos uint32, length int) {
+	var i uint32
+	for i = 0; i < uint32(length); i++ {
+		pb.addBase(refidx, startPos+i, dna.Gap)
+	}
+}
+
+// addInsertion to pileBuffer
+func addInsertion(pb *pileBuffer, refidx int, startPos uint32, seq []dna.Base) {
+	pb.addIns(refidx, startPos, dna.BasesToString(seq))
+}
+
+// ****** pileBuffer functions below ****** //
+
+// pileBuffer stores a slice of piles for efficient reuse.
+type pileBuffer struct {
+	s   []Pile // discontinuous positions are not allowed
```

Comment:
So many variables named "s", ha ha ha.  Is the "s" for slice here?

Reply:
yeahhhh, probably should change that. Changed to 'buf' for buffer